### ISSUE TITLE
[$225] Topics/Posts UI Updates

### ISSUE THREAD

**Issue by RishiRajSahu** [appirio-tech/connect-app#2180]
1.Identifying private topics on UI - we don't even have UI yet to mark topic as private
2.Identifying starred topics on UI
3.Showing count for unread messages for channels

Ref - https://github.com/appirio-tech/connect-app/issues/2112

Aha! Link: https://topcoder.aha.io/features/TCCONNECT-420

**Comment by maxceem** [comment]
@RishiRajSahu  As I understand for 1 and 2 we need to get backend updated, as we don't have yet the properties to mark topics as private or favorite.

**Comment by RishiRajSahu** [comment]
Designs for showing admins privileged topics - 

![image (1)](https://user-images.githubusercontent.com/11248478/54911347-4bbb0a00-4f14-11e9-8207-ed89dcf827e6.png)

**Comment by maxceem** [comment]
@RishiRajSahu Looks like there are no yet backend changes to support this feature. Let me know if you want to proceed with backend changes first. But I guess we would handle them outside of this CF.

**Comment by vikasrohit** [comment]
@maxceem we don't need any new backend support for this feature as we already have feature of non customer messages. It was an old feature where we used to show such discussions in a separate tab called `Discussions`. It is still visible for few of the old projects. What we need to do is to render those messages in the project dashboard with the new UI and only for non customers. We also need to provide ability to create such messages from the dashboard itself (May be we can add a simple checkbox for Copilots and Managers in the new thread component).

**Comment by vikasrohit** [comment]
Link to feature: https://tcprod.productboard.com/roadmap/251630-roadmap/features/1671343

**Comment by vikasrohit** [comment]
To unblock the issue @maxceem, we are going to have a simple checkbox (with label `Private`) just before the `Post` button. This would be visible for only topcoder team (copilots and managers) and selecting this checkbox would just change the `tag` of the topic to be posted as `SECONDARY` which are the posts we treat as non customer talks.
Also, we need to put the backend validation as well to prevent any one apart from the topcoder team to access `SECONDARY` tagged threads.

**Comment by maxceem** [comment]
Got it, thank you @vikasrohit 

1. For the old Discussion tab we currently use tag `MESSAGES`, should we keep using it, or use `SECONDARY` for private topics?

2. Should we somehow be able to switch `private`/`non-private` topic after creating it? 

3. Is it possible to get the source for locker icon for private posts?
    ![image](https://user-images.githubusercontent.com/146016/55699482-a3ba3c00-59fd-11e9-9cc3-82842bf558aa.png)

4. As I understand marking topics as a favorite is out of scope for now.

**Comment by RishiRajSahu** [comment]
@maxceem 
1.Okay, Let's use `MESSAGES` only.
2.No, as per me that won't be a case, even if it is in future we would be able to switch as its just the value of tag we have to change via PATCH
3.Sure, I will get it from vic.
4.Yes, for now it is out.

**Comment by maxceem** [comment]
Thank you @RishiRajSahu 

Do you have any preferences where to put a `Private` checkbox in the mobile view?

![image](https://user-images.githubusercontent.com/146016/55769985-4255a400-5ab5-11e9-890f-064bd7b84d08.png)

**Comment by RishiRajSahu** [comment]
@maxceem thanks for bringing it up, shall we ask in a small popup after hitting post button saying '**Is it a private thread ?**' for copilots and managers, what say ?

**Comment by maxceem** [comment]
@RishiRajSahu yes should work. As an alternative I can think is to add one more toolbar line on top of `Post` with a switcher `Private.
   
![image](https://user-images.githubusercontent.com/146016/55776402-32e35480-5acf-11e9-8190-c1b63e91c41c.png)

It would a little bit reduce the space for message text though.

**Comment by maxceem** [comment]
@RishiRajSahu when you have time, have look on this question. Which way do you prefer?

**Comment by RishiRajSahu** [comment]
@maxceem okay switch also works, are we using this same switch somewhere else in mobile view ?

**Comment by RishiRajSahu** [comment]
@maxceem its on listing page for admins, okay fine, let's use minimal best possible height for the switch and text div.

**Comment by maxceem** [comment]
@RishiRajSahu could you please confirm, if such design is good for us:

1. Wide screens:
    
    ![image](https://user-images.githubusercontent.com/146016/56110167-15f2c980-5f86-11e9-9fbc-b020f9867dbf.png)

2. Narrow screens:

     ![image](https://user-images.githubusercontent.com/146016/56110174-1db26e00-5f86-11e9-92fb-e5f942271101.png)

3. Mobile screens:

    ![image](https://user-images.githubusercontent.com/146016/56110188-286d0300-5f86-11e9-8d9a-e97c89c7084a.png)

    - For mobile screens, I suggest keeping the height for the panel with `Private` switch as `48px` as minimal suggested touch size is 44px in Apple’s iPhone Human Interface Guidelines and 48px in Android’s Material Design principles. To avoid users accidentally post a message when they want to switch to `Private` mode.

**Comment by maxceem** [comment]
Sum up:

1. Add a new permission to https://github.com/appirio-tech/connect-app/blob/dev/src/config/permissions.js#L11:
    ```
    ACCESS_PRIVATE_POST: {
      topcoderRoles: [PROJECT_ROLE_COPILOT, ROLE_CONNECT_MANAGER, ROLE_CONNECT_ADMIN, ROLE_ADMINISTRATOR],
    }
    ```

2. Show a `Private` switcher to the users which have `ACCESS_PRIVATE_POST` permission. You can use a method `checkPermission` to check permission. See usage example https://github.com/appirio-tech/connect-app/blob/dev/src/projects/detail/containers/ProjectPlanContainer.jsx#L172

3. When user switch on `Private`, when we send requests to the server to create a new Topic we should set `tag` as `MESSAGES` (https://github.com/appirio-tech/connect-app/blob/dev/src/config/constants.js#L607) instead of `PRIMARY` (https://github.com/appirio-tech/connect-app/blob/dev/src/config/constants.js#L606).

4. On the dashboard tab, we currently only show topics from the `PRIMARY` tag https://github.com/appirio-tech/connect-app/blob/dev/src/projects/actions/projectTopics.js#L29. For users who have `ACCESS_PRIVATE_POST` permission, we should additionally load from the server topics with `MESSAGES` tag. We already have an action for that which loads them into the Redux store: https://github.com/appirio-tech/connect-app/blob/dev/src/projects/actions/projectTopics.js#L40

    And after for such users, we should show these additional private topics under the regular topics 
    ![](https://user-images.githubusercontent.com/11248478/54911347-4bbb0a00-4f14-11e9-8207-ed89dcf827e6.png)

    - in the new UI we only have to make topics the same color and add a [locker icon](https://drive.google.com/open?id=1jNXHTgx0m3ygdgSgrLibh60uvPnP3Pgb) in front of the topic title. No other changes to the UI are required.

**Comment by maxceem** [comment]
The UI is described in the comment above https://github.com/appirio-tech/connect-app/issues/2180#issuecomment-483117786:

![image](https://user-images.githubusercontent.com/146016/56111626-383b1600-5f8b-11e9-900b-7608c2970f33.png)

![narr](https://user-images.githubusercontent.com/146016/56111740-92d47200-5f8b-11e9-932d-f20f9bfe9f13.png)

**Comment by maxceem** [comment]
This issues is now open for pick up.

**Comment by sumitdaga** [comment]
@maxceem  
can you please assign it to me?

**Comment by maxceem** [comment]
done

**Comment by sumitdaga** [comment]
@maxceem  
just a suggestion ...  why not move the switch in header  
or just below header instead of above the `Post` button , there is always a risk of posting by mistake when user is actually trying to toggle the switch   ..let me know ?

![Selection_140](https://user-images.githubusercontent.com/7769148/56115742-b85c7e00-5f81-11e9-86a2-567c09f20ca3.png)

**Comment by maxceem** [comment]
@sumitdaga great idea.

@RishiRajSahu what do you think?

Here is how it would look like on the minimal screen width `320px`:

Version 1:

![image](https://user-images.githubusercontent.com/146016/56116490-7be84c80-5f98-11e9-9f77-7e8b446a60ec.png)

Version 2:

![image](https://user-images.githubusercontent.com/146016/56116503-873b7800-5f98-11e9-860c-1578de62bc3c.png)

**Comment by RishiRajSahu** [comment]
@maxceem @sumitdaga version 2 looks good !

**Comment by sumitdaga** [comment]
@maxceem   
so we need to add a switcher only when creating a new post right? ...not to any other post (replies to post, phase discussion, etc) right ? only here   
![Selection_141](https://user-images.githubusercontent.com/7769148/56127421-bfdc5100-5f9a-11e9-8375-0adc0f8a0866.png)

**Comment by maxceem** [comment]
@sumitdaga right, only when we are creating a new topic as you showed with a red box above.

**Comment by sumitdaga** [comment]
@maxceem  
the `Discussions` section in the left sidebar i assume will not show these private message links  

btw this task is not as straightforward ...i would request you to review the prize once you check the PR!  
and i may need some more time as well
thanks

**Comment by maxceem** [comment]
> the Discussions section in the left sidebar i assume will not show these private message links

I think we can show links to private discussions to the users which may see them.

> btw this task is not as straightforward ...i would request you to review the prize once you check the PR!
> and i may need some more time as well

Sure, I can check it out during the review. Any particular difficulties you experience?

**Comment by sumitdaga** [comment]
@maxceem  
PR at #2957  
I could give you a detailed comment, but in short there were many places to check and verify (which is also clear from the number of files changed in PR) ..some things you should check:
- clearing `Private Switch` state when post is cancelled (by clicking outside or by cancel button)
- from the left sidebar if it scrolls to correct topic from both dashboard and project plan tab

let me know what you think after you check the PR and also if the implementation can be improved in some way!
thanks

**Comment by maxceem** [comment]
@sumitdaga 

I've reviewed the PR, great work has been done.
I'm going to raise the prize to $225 for this issue. Let me know what you think.

Thank you.

**Comment by sumitdaga** [comment]
@maxceem  
thanks much :)

**Comment by RishiRajSahu** [comment]
Working fine.
